### ISSUE TITLE
improve labeled graph terminology and documentation

### ISSUE THREAD

**Issue by johnyf** [tulip-control/tulip-control#18]
This issue is intended as a place to collect notable references on labeled graphs and fix a consistent terminology on automata, (state) transition systems and (state) machines. This post will be updated accordingly. Links:
- [Labeled MultiDigraphs](https://en.wikipedia.org/wiki/Labeled_multigraph): The abstraction to labeled graphs serves as a common base class (`tulip/transys/labeled_graphs`) providing facilities to define both discrete, as well as hybrid data structures, all of which are labeled graphs. Labeled graphs are already implemented by `networkx` and similar packages. Systems theory differs from pure labeled graphs primarily in that labeling functions may be [constrained](https://en.wikipedia.org/wiki/State_transition_system). For example the action labels annotating edges of a transition system (labeled edges are called transitions) must be elements from the action set of that transitions system. One purpose of `transys` is to provide methods for ensuring consistent labeling without too much coding overhead.
  Even subsets of states, e.g., the set of initial states, are just alternative representations of vertex labelings.
- [Automata and System Theory](http://www.csci.csusb.edu/dick/maths/math_71_Auto...Systems.html): interesting typing partially motivated by [Ross Ashby](https://en.wikipedia.org/wiki/William_Ross_Ashby)'s terminology, broadly in accord with `transys`.

The end result is not more documentation within the code, but a separate document on the subject and the confusion in the literature. Wikipedia features several cyclic definitions:
- [Automata](https://en.wikipedia.org/wiki/Automata_theory)
- [Semi-Automaton](https://en.wikipedia.org/wiki/Semiautomaton)
- [State Transition System](https://en.wikipedia.org/wiki/State_transition_system)
- [State Machine](https://en.wikipedia.org/wiki/State_Machine)
- [Abstract Machine](https://en.wikipedia.org/wiki/Abstract_machine)
- [State diagram](https://en.wikipedia.org/wiki/State_diagram)
- [Hybrid Automaton](https://en.wikipedia.org/wiki/Hybrid_automaton)
- [Timed Automaton](https://en.wikipedia.org/wiki/Timed_automaton)
- [I/O Automata](https://en.wikipedia.org/wiki/I/O_automaton)

Also, the term "state" may be redundant and interchangeable with (possibly labeled) vertex/node. Note that determinism, state finiteness, label set finiteness, and language finiteness are notions definable directly for labeled multidigraphs and not specialized to particular species. Only their interpretation varies.

Although in the literature transition systems have been named as automata (most notably "hybrid automata"), in general there is consensus on referring to automata as equipped with acceptance semantics in the form of some "accepting state" labeling.
- This agrees better with the etymology of [αυτόματα](https://en.wiktionary.org/wiki/automaton) as something which decides on its own, self-willed. In that sense, automata assign meaning to each input word as accepted or not.
- Transition systems describe possible output words (the output language defined by the possible sequences of actions). So although they are "self-willed", they define no acceptance semantics.
- Machines are input-output maps, so their execution is driven by their input. Note that [`Acacia+`](http://lit2.ulb.ac.be/acaciaplus/) clearly mentions that its output is a Moore machine (and not an automaton).

**Comment by slivingston** [comment]
@johnyf: do you still want to eventually change the names to `generators`, `acceptors`, and `transducers`? One alternative is to create a page in the User's Guide to include discussion about why different names may be desired and how to achieve it, e.g.,

```python
import tulip.transys.automata as acceptors
```

Obviously renaming imports is less than satisfactory because all of the documentation will continue to refer to `automata` etc. However, exposing users to the alternatives could be good preparation for an eventual switch.

Alternatively, we could mark this issue with a milestone like version 2.0 (or TuLiP 2000) so that it unambiguously refers to a big change in some possibly distant future.

**Comment by johnyf** [comment]
No, I do not intend to rename to `generators` and `acceptors` any more, because by now I consider these as automata (with different path quantification, existential for generators, and universal for the usual acceptor found in the literature, except for alternating automata). I would merge these two and substantially simplify, as done in [`omega.automata`](https://github.com/johnyf/omega/blob/0429586f8157cd22ac4c7eff84e7fe2e8facfc04/omega/automata.py).

I still think that `transducers` is more descriptive than `machines` (after all, they represent sequential transductions, so the terminology is in accord with the mathematical terminology). At a further level, transducers are again automata, with universal quantification (and the knowledge that, by construction, they satisfy the accepting condition), provided that the accepting condition be preserved in some form (not the case currently in `transys`).

However, I think that deferring such a change for the distant future (or not renaming at all) would be better. For example, I still try to represent the input to synthesis, as well as the result, with the class `omega.symbolic.symbolic.Automaton`, because it suffices, and helps against feature creep (and easier to remember). More or less, this is a refinement viewpoint, as (implicitly) advocated by [TLA](http://dx.doi.org/10.1145/177492.177726).
